File: back/src/main/java/com/plantmanager/service/HouseService.java
package com.plantmanager.service;

import com.plantmanager.dto.CreateHouseDTO;
import com.plantmanager.dto.HouseInvitationDTO;
import com.plantmanager.dto.HouseMemberDTO;
import com.plantmanager.dto.HouseResponseDTO;
import com.plantmanager.dto.JoinHouseDTO;
import com.plantmanager.entity.HouseEntity;
import com.plantmanager.entity.HouseInvitationEntity;
import com.plantmanager.entity.RoomEntity;
import com.plantmanager.entity.UserEntity;
import com.plantmanager.entity.UserHouseEntity;
import com.plantmanager.entity.NotificationEntity;
import com.plantmanager.entity.enums.InvitationStatus;
import com.plantmanager.entity.enums.NotificationType;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.ForbiddenException;
import jakarta.ws.rs.NotFoundException;
import com.plantmanager.entity.UserPlantEntity;
import com.plantmanager.entity.CareLogEntity;

import jakarta.inject.Inject;

import org.hibernate.Hibernate;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Service for house business logic.
 */
@ApplicationScoped
public class HouseService {

    @Inject
    FcmService fcmService;

    @Inject
    GamificationService gamificationService;

    /**
     * Get all houses for a user.
     */
    public List<HouseResponseDTO> getUserHouses(UUID userId) {
        List<UserHouseEntity> memberships = UserHouseEntity.findByUser(userId);
        return memberships.stream()
                .map(m -> HouseResponseDTO.from(m.house, m))
                .toList();
    }

    /**
     * Get the active house for a user.
     */
    public HouseResponseDTO getActiveHouse(UUID userId) {
        UserHouseEntity membership = UserHouseEntity.findActiveByUser(userId);
        if (membership == null) {
            throw new NotFoundException("No active house found");
        }
        return HouseResponseDTO.from(membership.house, membership);
    }

    /**
     * Get a house by ID (with membership check).
     */
    public HouseResponseDTO getHouseById(UUID userId, UUID houseId) {
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (membership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }
        return HouseResponseDTO.from(membership.house, membership);
    }

    /**
     * Create a new house (user becomes OWNER).
     */
    @Transactional
    public HouseResponseDTO createHouse(UUID userId, CreateHouseDTO dto) {
        UserEntity user = UserEntity.findById(userId);
        if (user == null) {
            throw new NotFoundException("User not found");
        }

        // Create the house
        HouseEntity house = new HouseEntity();
        house.name = dto.name();
        house.persist();

        // Create membership (as OWNER and active)
        UserHouseEntity membership = new UserHouseEntity();
        membership.user = user;
        membership.house = house;
        membership.role = UserEntity.UserRole.OWNER;
        membership.isActive = true;
        membership.persist();

        // Deactivate other houses for this user
        deactivateOtherHouses(userId, house.id);

        return HouseResponseDTO.from(house, membership);
    }

    /**
     * Request to join a house using invite code.
     * Creates a PENDING invitation that must be accepted by an owner.
     */
    @Transactional
    public HouseInvitationDTO requestJoinHouse(UUID userId, JoinHouseDTO dto) {
        UserEntity user = UserEntity.findById(userId);
        if (user == null) {
            throw new NotFoundException("User not found");
        }

        HouseEntity house = HouseEntity.findByInviteCode(dto.inviteCode())
                .orElseThrow(() -> new NotFoundException("Invalid invite code"));

        // Check if already a member
        UserHouseEntity existing = UserHouseEntity.findByUserAndHouse(userId, house.id);
        if (existing != null) {
            throw new BadRequestException("Vous etes deja membre de cette maison");
        }

        // Check if already has a pending request
        HouseInvitationEntity existingInvitation = HouseInvitationEntity.findPendingByRequesterAndHouse(userId, house.id);
        if (existingInvitation != null) {
            throw new BadRequestException("Vous avez deja une demande en attente pour cette maison");
        }

        // Create pending invitation
        HouseInvitationEntity invitation = new HouseInvitationEntity();
        invitation.house = house;
        invitation.requester = user;
        invitation.status = InvitationStatus.PENDING;
        invitation.persist();

        // Send notification to all owners of the house
        List<UserHouseEntity> owners = UserHouseEntity.list("house.id = ?1 and role = ?2",
                house.id, UserEntity.UserRole.OWNER);
        for (UserHouseEntity owner : owners) {
            NotificationEntity notification = new NotificationEntity();
            notification.user = owner.user;
            notification.type = NotificationType.HOUSE_INVITATION;
            notification.message = user.displayName + " souhaite rejoindre votre maison \"" + house.name + "\"";
            notification.invitation = invitation;
            notification.house = house;
            notification.persist();
        }

        // Send push notification to owners
        fcmService.sendToHouseMembers(
                house.id,
                userId,
                "Demande d'adhesion",
                user.displayName + " souhaite rejoindre votre maison",
                Map.of("type", "HOUSE_INVITATION", "houseId", house.id.toString(), "invitationId", invitation.id.toString())
        );

        return HouseInvitationDTO.from(invitation);
    }

    /**
     * Accept a join request (OWNER only).
     */
    @Transactional
    public HouseInvitationDTO acceptInvitation(UUID ownerId, UUID invitationId) {
        HouseInvitationEntity invitation = HouseInvitationEntity.findById(invitationId);
        if (invitation == null) {
            throw new NotFoundException("Demande introuvable");
        }

        // Verify the responder is an OWNER of the house
        UserHouseEntity ownerMembership = UserHouseEntity.findByUserAndHouse(ownerId, invitation.house.id);
        if (ownerMembership == null || ownerMembership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Seul le proprietaire peut accepter les demandes");
        }

        if (invitation.status != InvitationStatus.PENDING) {
            throw new BadRequestException("Cette demande a deja ete traitee");
        }

        UserEntity owner = UserEntity.findById(ownerId);

        // Accept the invitation
        invitation.status = InvitationStatus.ACCEPTED;
        invitation.respondedBy = owner;
        invitation.respondedAt = OffsetDateTime.now();

        // Create membership for the requester
        UserHouseEntity membership = new UserHouseEntity();
        membership.user = invitation.requester;
        membership.house = invitation.house;
        membership.role = UserEntity.UserRole.MEMBER;
        membership.isActive = true;
        membership.persist();

        // Deactivate other houses for the new member
        deactivateOtherHouses(invitation.requester.id, invitation.house.id);

        // Gamification
        gamificationService.onHouseJoined(invitation.requester.id, invitation.house.id);

        // Notify the requester that they were accepted
        NotificationEntity notification = new NotificationEntity();
        notification.user = invitation.requester;
        notification.type = NotificationType.MEMBER_JOINED;
        notification.message = "Votre demande pour rejoindre \"" + invitation.house.name + "\" a ete acceptee !";
        notification.house = invitation.house;
        notification.persist();

        // Push notification to the requester
        fcmService.sendToUser(
                invitation.requester.id,
                "Demande acceptee",
                "Vous avez rejoint la maison \"" + invitation.house.name + "\"",
                Map.of("type", "MEMBER_JOINED", "houseId", invitation.house.id.toString())
        );

        // Notify existing members
        fcmService.sendToHouseMembers(
                invitation.house.id,
                invitation.requester.id,
                "Nouveau membre",
                invitation.requester.displayName + " a rejoint la maison",
                Map.of("type", "MEMBER_JOINED", "houseId", invitation.house.id.toString())
        );

        return HouseInvitationDTO.from(invitation);
    }

    /**
     * Decline a join request (OWNER only).
     */
    @Transactional
    public HouseInvitationDTO declineInvitation(UUID ownerId, UUID invitationId) {
        HouseInvitationEntity invitation = HouseInvitationEntity.findById(invitationId);
        if (invitation == null) {
            throw new NotFoundException("Demande introuvable");
        }

        // Verify the responder is an OWNER of the house
        UserHouseEntity ownerMembership = UserHouseEntity.findByUserAndHouse(ownerId, invitation.house.id);
        if (ownerMembership == null || ownerMembership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Seul le proprietaire peut refuser les demandes");
        }

        if (invitation.status != InvitationStatus.PENDING) {
            throw new BadRequestException("Cette demande a deja ete traitee");
        }

        UserEntity owner = UserEntity.findById(ownerId);

        // Decline the invitation
        invitation.status = InvitationStatus.DECLINED;
        invitation.respondedBy = owner;
        invitation.respondedAt = OffsetDateTime.now();

        // Notify the requester
        NotificationEntity notification = new NotificationEntity();
        notification.user = invitation.requester;
        notification.type = NotificationType.HOUSE_INVITATION;
        notification.message = "Votre demande pour rejoindre \"" + invitation.house.name + "\" a ete refusee.";
        notification.house = invitation.house;
        notification.persist();

        // Push notification to the requester
        fcmService.sendToUser(
                invitation.requester.id,
                "Demande refusee",
                "Votre demande pour \"" + invitation.house.name + "\" a ete refusee",
                Map.of("type", "HOUSE_INVITATION", "houseId", invitation.house.id.toString())
        );

        return HouseInvitationDTO.from(invitation);
    }

    /**
     * Get pending invitations for a house (OWNER only).
     */
    public List<HouseInvitationDTO> getPendingInvitations(UUID ownerId, UUID houseId) {
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(ownerId, houseId);
        if (membership == null || membership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Seul le proprietaire peut voir les demandes");
        }

        List<HouseInvitationEntity> invitations = HouseInvitationEntity.findPendingByHouse(houseId);
        return invitations.stream()
                .map(HouseInvitationDTO::from)
                .toList();
    }

    /**
     * Get my pending join requests (as requester).
     */
    public List<HouseInvitationDTO> getMyPendingRequests(UUID userId) {
        List<HouseInvitationEntity> invitations = HouseInvitationEntity.findPendingByRequester(userId);
        return invitations.stream()
                .map(HouseInvitationDTO::from)
                .toList();
    }

    /**
     * Switch active house.
     */
    @Transactional
    public HouseResponseDTO switchActiveHouse(UUID userId, UUID houseId) {
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (membership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }

        // Deactivate all other houses
        deactivateOtherHouses(userId, houseId);

        // Activate this house
        membership.isActive = true;

        return HouseResponseDTO.from(membership.house, membership);
    }

    /**
     * Leave a house.
     */
    @Transactional
    public void leaveHouse(UUID userId, UUID houseId) {
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (membership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }

        // Can't leave if you're the only OWNER
        if (membership.role == UserEntity.UserRole.OWNER) {
            long ownerCount = UserHouseEntity.count("house.id = ?1 and role = ?2",
                    houseId, UserEntity.UserRole.OWNER);
            if (ownerCount <= 1) {
                throw new BadRequestException("Cannot leave: you are the only owner. Transfer ownership first.");
            }
        }

        boolean wasActive = membership.isActive;
        membership.delete();

        // If this was the active house, activate another one
        if (wasActive) {
            List<UserHouseEntity> remaining = UserHouseEntity.findByUser(userId);
            if (!remaining.isEmpty()) {
                remaining.get(0).isActive = true;
            }
        }
    }

    /**
     * Delete a house (Owner only).
     */
    @Transactional
    public void deleteHouse(UUID userId, UUID houseId) {
        // Fetch the house (managed entity)
        HouseEntity house = HouseEntity.findById(houseId);
        if (house == null) {
            throw new NotFoundException("House not found");
        }

        // Verify membership and ownership
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (membership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }

        if (membership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Only the owner can delete the house");
        }

        // 1. Manually delete all rooms (and their plants/children)
        // We iterate specifically to trigger cascades and avoid "TransientPropertyValueException"
        List<RoomEntity> rooms = RoomEntity.findByHouse(houseId);
        for (RoomEntity room : rooms) {
            // Delete plants in this room
            List<UserPlantEntity> plants = UserPlantEntity.findByRoom(room.id);
            for (UserPlantEntity plant : plants) {
                // This deletes logs and notifications via cascade if configured,
                // otherwise we might need to delete them explicitly too.
                // Given previous errors, let's be safe:
                CareLogEntity.delete("plant.id = ?1", plant.id);
                NotificationEntity.delete("plant.id = ?1", plant.id);
                plant.delete();
            }
            room.delete();
        }
        
        // 2. Delete all memberships manually
        List<UserHouseEntity> members = UserHouseEntity.list("house.id", houseId);
        for (UserHouseEntity m : members) {
            m.delete();
        }

        // 3. Finally delete the house
        house.delete();
    }

    // ==================== MEMBER MANAGEMENT ====================

    /**
     * Get all members of a house (must be a member to view).
     */
    public List<HouseMemberDTO> getHouseMembers(UUID userId, UUID houseId) {
        // Verify requester is a member
        UserHouseEntity requesterMembership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (requesterMembership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }

        List<UserHouseEntity> members = UserHouseEntity.findByHouse(houseId);
        return members.stream()
                .map(HouseMemberDTO::from)
                .toList();
    }

    /**
     * Update a member's role (OWNER only).
     * Can promote MEMBER to OWNER or demote OWNER to MEMBER.
     */
    @Transactional
    public HouseMemberDTO updateMemberRole(UUID requesterId, UUID houseId, UUID targetUserId, String newRole) {
        // Verify requester is OWNER
        UserHouseEntity requesterMembership = UserHouseEntity.findByUserAndHouse(requesterId, houseId);
        if (requesterMembership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }
        if (requesterMembership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Only owners can change member roles");
        }

        // Find target member
        UserHouseEntity targetMembership = UserHouseEntity.findByUserAndHouse(targetUserId, houseId);
        if (targetMembership == null) {
            throw new NotFoundException("Member not found in this house");
        }

        // Parse and validate new role
        UserEntity.UserRole role;
        try {
            role = UserEntity.UserRole.valueOf(newRole.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BadRequestException("Invalid role. Must be OWNER, MEMBER or GUEST");
        }

        // Prevent demoting yourself if you're the only OWNER
        if (requesterId.equals(targetUserId) && role != UserEntity.UserRole.OWNER) {
            long ownerCount = UserHouseEntity.count("house.id = ?1 and role = ?2",
                    houseId, UserEntity.UserRole.OWNER);
            if (ownerCount <= 1) {
                throw new BadRequestException("Cannot demote: you are the only owner. Promote another member first.");
            }
        }

        // Update role
        targetMembership.role = role;

        return HouseMemberDTO.from(targetMembership);
    }

    /**
     * Remove a member from a house (OWNER only).
     * Cannot remove yourself (use leaveHouse instead).
     */
    @Transactional
    public void removeMember(UUID requesterId, UUID houseId, UUID targetUserId) {
        // Can't kick yourself
        if (requesterId.equals(targetUserId)) {
            throw new BadRequestException("Cannot remove yourself. Use leave house instead.");
        }

        // Verify requester is OWNER
        UserHouseEntity requesterMembership = UserHouseEntity.findByUserAndHouse(requesterId, houseId);
        if (requesterMembership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }
        if (requesterMembership.role != UserEntity.UserRole.OWNER) {
            throw new ForbiddenException("Only owners can remove members");
        }

        // Find target member
        UserHouseEntity targetMembership = UserHouseEntity.findByUserAndHouse(targetUserId, houseId);
        if (targetMembership == null) {
            throw new NotFoundException("Member not found in this house");
        }

        // Delete membership
        targetMembership.delete();

        // If target had this as active house, activate another one for them
        List<UserHouseEntity> remaining = UserHouseEntity.findByUser(targetUserId);
        if (!remaining.isEmpty() && remaining.stream().noneMatch(m -> m.isActive)) {
            remaining.get(0).isActive = true;
        }
    }

    /**
     * Helper to deactivate all houses for a user except one.
     */
    private void deactivateOtherHouses(UUID userId, UUID exceptHouseId) {
        UserHouseEntity.update("isActive = false where user.id = ?1 and house.id != ?2",
                userId, exceptHouseId);
    }

    /**
     * Get house-wide activity feed: recent care logs from all members.
     */
    public List<com.plantmanager.dto.CareLogDTO> getHouseActivity(UUID userId, UUID houseId, int limit) {
        // Verify membership
        UserHouseEntity membership = UserHouseEntity.findByUserAndHouse(userId, houseId);
        if (membership == null) {
            throw new ForbiddenException("You are not a member of this house");
        }

        List<CareLogEntity> logs = CareLogEntity.findByHouse(houseId, limit);

        // Force load lazy fields
        return logs.stream()
                .peek(log -> {
                    if (log.plant != null) Hibernate.initialize(log.plant);
                    if (log.user != null) Hibernate.initialize(log.user);
                })
                .map(com.plantmanager.dto.CareLogDTO::from)
                .toList();
    }
}

File: front/lib/core/services/house_service.dart
import 'package:dio/dio.dart';
import 'package:planto/core/models/house.dart';
import 'package:planto/core/models/house_invitation.dart';
import 'package:planto/core/models/house_member.dart';
import 'package:planto/core/models/vacation_delegation.dart';
import 'package:planto/core/services/api_client.dart';
import 'package:planto/core/services/cache_service.dart';
import 'package:planto/core/utils/api_error_formatter.dart';

/// Service for house API operations.
/// Uses ApiClient with automatic token refresh.
/// Supporte le mode hors-ligne via cache local.
class HouseService {
  late final Dio _dio;
  final CacheService _cache = CacheService.instance;

  /// Indique si la dernière requête a été servie depuis le cache
  bool lastRequestFromCache = false;

  HouseService({Dio? dio}) : _dio = dio ?? ApiClient.instance.dio;

  /// Get all houses for the current user
  Future<List<House>> getMyHouses() async {
    const cacheKey = 'my_houses';
    lastRequestFromCache = false;

    try {
      final response = await _dio.get('/api/v1/houses');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        await _cache.putList(cacheKey, data);
        return data.map((json) => House.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load houses');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 401) {
        throw Exception('Session expirée');
      }
      // Mode hors-ligne
      final cached = await _cache.getList(cacheKey);
      if (cached != null) {
        lastRequestFromCache = true;
        return cached
            .map((json) => House.fromJson(json as Map<String, dynamic>))
            .toList();
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de charger vos maisons'),
      );
    }
  }

  /// Get the active house
  Future<House> getActiveHouse() async {
    try {
      final response = await _dio.get('/api/v1/houses/active');

      if (response.statusCode == 200) {
        return House.fromJson(response.data);
      } else {
        throw Exception('No active house');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 404) {
        throw Exception('Aucune maison active');
      }
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger la maison active',
        ),
      );
    }
  }

  /// Switch active house
  Future<House> switchActiveHouse(String houseId) async {
    try {
      final response = await _dio.put('/api/v1/houses/$houseId/activate');

      if (response.statusCode == 200) {
        return House.fromJson(response.data);
      } else {
        throw Exception('Failed to switch house');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        throw Exception('Vous n\'etes pas membre de cette maison');
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de changer de maison'),
      );
    }
  }

  /// Create a new house
  Future<House> createHouse(String name) async {
    try {
      final response = await _dio.post('/api/v1/houses', data: {'name': name});

      if (response.statusCode == 201) {
        return House.fromJson(response.data);
      } else {
        throw Exception('Failed to create house');
      }
    } on DioException catch (e) {
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de creer la maison'),
      );
    }
  }

  /// Request to join a house with invite code (creates a pending request)
  Future<HouseInvitation> requestJoinHouse(String inviteCode) async {
    try {
      final response = await _dio.post(
        '/api/v1/houses/join',
        data: {'inviteCode': inviteCode},
      );

      if (response.statusCode == 201) {
        return HouseInvitation.fromJson(response.data);
      } else {
        throw Exception('Failed to send join request');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 404) {
        throw Exception('Code d\'invitation invalide');
      }
      if (e.response?.statusCode == 400) {
        throw Exception(
          formatApiError(
            e,
            fallbackMessage: 'Impossible d\'envoyer la demande',
          ),
        );
      }
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible d\'envoyer la demande',
        ),
      );
    }
  }

  /// Accept a join request (Owner only)
  Future<HouseInvitation> acceptInvitation(String invitationId) async {
    try {
      final response = await _dio.put(
        '/api/v1/houses/invitations/$invitationId/accept',
      );

      if (response.statusCode == 200) {
        return HouseInvitation.fromJson(response.data);
      } else {
        throw Exception('Failed to accept invitation');
      }
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible d\'accepter la demande',
        ),
      );
    }
  }

  /// Decline a join request (Owner only)
  Future<HouseInvitation> declineInvitation(String invitationId) async {
    try {
      final response = await _dio.put(
        '/api/v1/houses/invitations/$invitationId/decline',
      );

      if (response.statusCode == 200) {
        return HouseInvitation.fromJson(response.data);
      } else {
        throw Exception('Failed to decline invitation');
      }
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de refuser la demande',
        ),
      );
    }
  }

  /// Get pending invitations for a house (Owner only)
  Future<List<HouseInvitation>> getPendingInvitations(String houseId) async {
    try {
      final response = await _dio.get('/api/v1/houses/$houseId/invitations');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => HouseInvitation.fromJson(json)).toList();
      }
      return [];
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger les demandes',
        ),
      );
    }
  }

  /// Get my pending join requests
  Future<List<HouseInvitation>> getMyPendingRequests() async {
    try {
      final response = await _dio.get('/api/v1/houses/my-requests');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => HouseInvitation.fromJson(json)).toList();
      }
      return [];
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger vos demandes',
        ),
      );
    }
  }

  /// Leave a house
  Future<void> leaveHouse(String houseId) async {
    try {
      await _dio.delete('/api/v1/houses/$houseId/leave');
    } on DioException catch (e) {
      if (e.response?.statusCode == 400) {
        throw Exception(
          formatApiError(
            e,
            fallbackMessage: 'Impossible de quitter cette maison',
          ),
        );
      }
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de quitter cette maison',
        ),
      );
    }
  }

  /// Delete a house (Owner only)
  Future<void> deleteHouse(String houseId) async {
    try {
      await _dio.delete('/api/v1/houses/$houseId');
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        throw Exception('Seul le proprietaire peut supprimer la maison');
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de supprimer la maison'),
      );
    }
  }

  // ==================== MEMBER MANAGEMENT ====================

  /// Get all members of a house
  Future<List<HouseMember>> getHouseMembers(String houseId) async {
    try {
      final response = await _dio.get('/api/v1/houses/$houseId/members');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => HouseMember.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load members');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        throw Exception('Vous n\'etes pas membre de cette maison');
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de charger les membres'),
      );
    }
  }

  /// Update a member's role (Owner only)
  Future<HouseMember> updateMemberRole(
    String houseId,
    String userId,
    String newRole,
  ) async {
    try {
      final response = await _dio.put(
        '/api/v1/houses/$houseId/members/$userId/role',
        data: {'role': newRole},
      );

      if (response.statusCode == 200) {
        return HouseMember.fromJson(response.data);
      } else {
        throw Exception('Failed to update role');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        throw Exception('Seul le proprietaire peut modifier les roles');
      }
      if (e.response?.statusCode == 400) {
        throw Exception(
          formatApiError(e, fallbackMessage: 'Action impossible'),
        );
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible de modifier le role'),
      );
    }
  }

  /// Remove a member from a house (Owner only)
  Future<void> removeMember(String houseId, String userId) async {
    try {
      await _dio.delete('/api/v1/houses/$houseId/members/$userId');
    } on DioException catch (e) {
      if (e.response?.statusCode == 403) {
        throw Exception('Seul le proprietaire peut exclure des membres');
      }
      if (e.response?.statusCode == 400) {
        throw Exception(
          formatApiError(e, fallbackMessage: 'Action impossible'),
        );
      }
      throw Exception(
        formatApiError(e, fallbackMessage: 'Impossible d\'exclure ce membre'),
      );
    }
  }

  // ==================== VACATION / DELEGATION ====================

  /// Activate vacation mode (delegate plant care to another member)
  Future<VacationDelegation> activateVacation(
    String houseId, {
    required String delegateId,
    required String startDate,
    required String endDate,
    String? message,
  }) async {
    try {
      final data = <String, dynamic>{
        'delegateId': delegateId,
        'startDate': startDate,
        'endDate': endDate,
      };
      if (message != null && message.isNotEmpty) {
        data['message'] = message;
      }

      final response = await _dio.post(
        '/api/v1/houses/$houseId/vacation',
        data: data,
      );

      if (response.statusCode == 201) {
        return VacationDelegation.fromJson(response.data);
      } else {
        throw Exception('Impossible d\'activer le mode vacances');
      }
    } on DioException catch (e) {
      if (e.response?.statusCode == 400) {
        throw Exception(formatApiError(e, fallbackMessage: 'Requete invalide'));
      }
      if (e.response?.statusCode == 403) {
        throw Exception('Vous n\'etes pas membre de cette maison');
      }
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible d\'activer le mode vacances',
        ),
      );
    }
  }

  /// Cancel active vacation
  Future<void> cancelVacation(String houseId) async {
    try {
      await _dio.delete('/api/v1/houses/$houseId/vacation');
    } on DioException catch (e) {
      if (e.response?.statusCode == 404) {
        throw Exception('Aucune delegation active');
      }
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible d\'annuler le mode vacances',
        ),
      );
    }
  }

  /// Get current vacation status (returns null if no active vacation)
  Future<VacationDelegation?> getVacationStatus(String houseId) async {
    try {
      final response = await _dio.get('/api/v1/houses/$houseId/vacation');

      if (response.statusCode == 200 && response.data != null) {
        return VacationDelegation.fromJson(response.data);
      }
      return null; // 204 No Content
    } on DioException catch (e) {
      if (e.response?.statusCode == 204) return null;
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger le mode vacances',
        ),
      );
    }
  }

  /// Get all active delegations in the house
  Future<List<VacationDelegation>> getHouseDelegations(String houseId) async {
    try {
      final response = await _dio.get('/api/v1/houses/$houseId/delegations');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => VacationDelegation.fromJson(json)).toList();
      }
      return [];
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger les delegations',
        ),
      );
    }
  }

  /// Get delegations where I am the delegate
  Future<List<VacationDelegation>> getMyDelegations(String houseId) async {
    try {
      final response = await _dio.get('/api/v1/houses/$houseId/my-delegations');

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data;
        return data.map((json) => VacationDelegation.fromJson(json)).toList();
      }
      return [];
    } on DioException catch (e) {
      throw Exception(
        formatApiError(
          e,
          fallbackMessage: 'Impossible de charger vos delegations',
        ),
      );
    }
  }
}
